csi.proto

```diff
@@ -271,15 +271,18 @@ message VolumeMetadata {
 }
 
 // A standard way to encode credential data. The total bytes of the values in
-// the Data field must be less than MaxSecretSize bytes.
+// the Data field must be less than 1 Mebibyte.
 message Credentials {
-  // Data contains the credential data, for example username and password. Each
-  // key must consist of alphanumeric characters, '-', '_' or '.'. The
-  // serialized form of the secret data is a base64 encoded string, representing
-  // the arbitrary (possibly non-string) data value here.
-  // This information is sensitive and should be treated as such (not logged,
-  // etc.)
-  map<string, bytes> Data = 1;
+  // Data contains the credential data, for example username and password.
+  // Each key must consist of alphanumeric characters, '-', '_' or '.'.
+  // Each value MUST contain a valid string. An SP MAY choose to accept binary
+  // (non-string) data by using a binary-to-text encoding scheme, like base64.
+  // An SP SHALL advertise the requirements for credentials in documentation.
+  // COs SHALL permit users to pass through the required credentials.
+  // This information is sensitive and MUST be treated as such (not logged,
+  // etc.) by the CO.
+  // This field is REQUIRED.
+  map<string, string> Data = 1;
```

Comment:
> How will the CO know if the value is base64-encoded binary or regular UTF-8 encoded text?

The CO will not need to know. As far as the CO is concerned it needs to find a populate a *string* field. Only SP and end user/cluster admin will need to be aware that binary data needs to be passed in for a particular key, and as part of the documentation the SP will specify which binary to text encoding to use (base64 or quoted printable), and when handing over that secret to the CO the end user/cluster admin will ensure the correct encoding is applied and the CO gets string data.

> Do you perhaps intend that for a given key (eg., "password"), the value must always be UTF-8 encoded text?

Nope. UTF-8 is a character encoding, if we operate directly on strings, instead of byte arrays, senders don't have to worry about the character encoding, since the underlying protocol, gRPC, takes care of that (serializing/deserializing the string over the wire).

> I prefer encoding text to UTF-8 on it's way to another process over encoding binary to base64.

Because in the majority of cases this will likely be used to transfer string data (username, password, etc.), we should simplify that and not requiring callers to have to navigate encoding (make it dead simple: you have a string, give me a string, don't have to worry about anything else since gRPC takes care of transport).